File: android/src/main/kotlin/com/example/xplayer/XplayerPlugin.kt
package com.example.xplayer

import android.content.Context
import com.example.xplayer.views.XPlayerViewFactory
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.platform.PlatformViewRegistry

/** XplayerPlugin */
class XplayerPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private lateinit var platformRegistry: PlatformViewRegistry

    private val xplayer = XPlayer()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "xplayer")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "xplayer_events:state")
        eventChannel.setStreamHandler(xplayer)

        platformRegistry = flutterPluginBinding.platformViewRegistry
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "xplayer_viewer_default",
            XPlayerViewFactory(xplayer),
        )

    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "xplayer:removeView" -> xplayer.playerViewController.removeView(call) //d

            "xplayer:init" -> xplayer.init(context, result) //d
            "xplayer:claimPlayer" -> xplayer.claimPlayer(call) //d
            "xplayer:registerPlaylist" -> xplayer.registerPlaylist(call, result)
            "xplayer:changePlaylist" -> xplayer.changePlaylist(call, result)
            "xplayer:seekToNext" -> xplayer.seekToNext() //d
            "xplayer:seekTo" -> xplayer.seekTo(call) //d
            "xplayer:setPlayBackSpeed" -> xplayer.setPlayBackSpeed(call) //d
            "xplayer:seekToPreviousMediaItem" -> xplayer.seekToPreviousMediaItem() //d
            "xplayer:addMediaSource" -> xplayer.addMediaSource(call) //d
            "xplayer:addMediaSources" -> xplayer.addMediaSources(call, result) //d
            "xplayer:setMediaSource" -> xplayer.setMediaSource(call) //d
            "xplayer:setMediaSources" -> xplayer.setMediaSources(call, result) //d
            "xplayer:changeQuality" -> xplayer.changeQuality(call) //d
            "xplayer:clearAllMediaSource" -> xplayer.clearAllMediaSource() //d
            "xplayer:play" -> xplayer.play() //d
            "xplayer:pause" -> xplayer.pause() //d
            "xplayer:dispose" -> xplayer.dispose(result) //d
            else -> result.success("Method Channel Not Available")
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        xplayer.dispose(null)
    }

}

File: ios/Classes/XplayerPlugin.swift
import Flutter
import UIKit

public class XplayerPlugin: NSObject, FlutterPlugin {
    let xplayer: XPlayer = XPlayer()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = XplayerPlugin()
        
        let channel = FlutterMethodChannel(name: "xplayer", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        let eventChannel = FlutterEventChannel(name: "xplayer_events:state", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance.xplayer)
        
        
        let xplayerViewFactory = XPlayerViewFactory(xplayer: instance.xplayer, messenger: registrar.messenger())
        registrar.register(xplayerViewFactory, withId: "xplayer_viewer_default")
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        print("CallMethod" + call.method)
        switch call.method {
        case "getPlatformVersion":
            result("iOS fdfd" + UIDevice.current.systemVersion)
            
        case "xplayer:init":
            xplayer.initilize(result: result)
            
        case "xplayer:claimPlayer":
            xplayer.claimPlayer(call: call, result: result)
            
        case "xplayer:registerPlaylist":
            xplayer.registerPlaylist(call: call, result: result)
            
        case "xplayer:changePlaylist":
            xplayer.changePlaylist(call: call, result: result)
            
        case "xplayer:seekToNext":
            xplayer.seekToNext(call: call, result: result)
            
        case "xplayer:seekTo":
            xplayer.seekTo()
            result("seekTo")
            
        case "xplayer:seekToPreviousMediaItem":
            xplayer.seekToPreviousMediaItem(call: call, result: result)
            
        case "xplayer:setPlayBackSpeed":
            xplayer.setPlayBackSpeed()
            result("setPlayBackSpeed")
            
        case "xplayer:addMediaSource":
            xplayer.addMediaSource(call: call, result: result)
            result("addMediaSource")
            
        case "xplayer:addMediaSources":
            xplayer.addMediaSources(call: call, result: result)
            result("addMediaSources")
            
        case "xplayer:setMediaSource":
            xplayer.setMediaSource(call: call, result: result)
            result("setMediaSource")
            
        case "xplayer:setMediaSources":
            xplayer.setMediaSources(call: call, result: result)
            result("setMediaSources")
            
        case "xplayer:clearMediaSource":
            xplayer.clearMediaSource()
            result("clearMediaSource")
            
        case "xplayer:changeQuality":
            xplayer.changeQuality()
            result("changeQuality")
            
        case "xplayer:clearAllMediaSource":
            xplayer.clearAllMediaSource()
            result("clearAllMediaSource")
            
        case "xplayer:play":
            xplayer.play()
            result("play")
            
        case "xplayer:pause":
            xplayer.pause()
            result("pause")
            
        case "xplayer:dispose":
            xplayer.dispose()
            result("dispose")
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
}
